### PR TITLE
Timeline widget

### PR DESCRIPTION
**Pull request by r37r05p3C7**
Simple timeline keeping track of different events.
Here is a widget itself and all currently available events:

image

Full timestamp can be seen on hover(formatted using `globals.settings.timestamp_format`):

I'll think about more possible events and promote to pr from draft later.

Also, let me know if you think `TimelineEvent` can be stored in a better way. I don't think `sql_to_py` can handle nested objects so  I just used `StrEnum` and some pattern matching to work around that.

### CODE REVIEW

**Comment by Willy-JL** [comment]
I'm really liking this idea. Needs some work (for example launching could make a bit too many events), and I'll have a look at the code when I have the time, but deffo going in a good direction!

**Comment by Willy-JL** [comment]
- indeed, this structure would not work for `sql_to_py()`. however, i dont see why a normal int enum would not work. you can match `event_type` (`int`) to `TimelineEvent.GameInstalled.value` for example. not sure if you would need `event_type.value` when appending tho, i think not because json should try to make it into an int (intenum is a subclass of int)
- storing the textual description makes it harder to update in the future if translations ever get considered or if the message needs to be changed. could just store the parameters as a tuple. can take a variable amount of arguments as `add_timeline_event(self, event_type: TimelineEvent, *args)` and do `.append((time, event_type, args))`. in structs it could be `timeline_events: list[tuple[int, TimelineEvent, tuple]]` leaving that last tuple with no specification since it depends on the event type. then for example `game.add_timeline_event(TimelineEvent.ChangedName, old_name, new_name)`, and draw would expect the tuple to have a known amount of elements and types to make the message on the fly. ACTUALLY, even better, can make a (data?)class for `TimelineEvent` (and rename the existing one to `TimelineEventType`, or better move it inside the new one so its `TimelineEvent.Type`) which takes and stores the timestamp (as a `Timestamp` object, so the textual representation is available via `.display`), the type (as the datatype explained before) and the variable list of arguments, which would be taken into consideration in the `display()` property of this new `TimelineEvent`, so then you can just use `event.display` and it gets a string representation based on the data it has.
- drawing should do as little operations as possible. sorting and creating a new list is intensive. rather, the events should always be sorted (added in the correct order) then

**Comment by Willy-JL** [comment]
actually now i kinda wanna try doing these changes myself, sounds like fun lol

is there anything else you would we waiting on besides new events and what i said in order to undraft?

**Comment by FaceCrap** [comment]
Looks like something useful. Will come in very handy to keep track of update cycles (as in how often does a game get updated)

On comment though... I dislike wasted space... and the image in top shows a lot of that. Since you're forcing a sequential order already, the connection lines between the icons seems redundant. Losing that line would make the display a lot more compact.

**Comment by Willy-JL** [comment]
> I think I'll place a dropdown within the "Timeline" tab itself, positioned right at the top for easy access. Do you think it's a good idea to persist timeline settings?

yes and yes, imo



> `GameAdded` event already uses `added_on` field. It will be harder to retroactively add other fields like `last_updated`, `last_played`, etc. because they are not static. To include them in the timeline, we'd have to do a small database migration on startup, which I don't think is worth it.

agreed

> How about this layout? I prefer it because we won't have to deal with inconsistent, weird looking monospace fonts.

i like it

**Comment by FaceCrap** [comment]
> Do you think it's a good idea to persist timeline settings?

Yes!. By all means Yes! That is why I suggested to also add the same dropdown in the sidebar. But, if the selection made in the tab itself is persistent, the sidebar option would not be needed. As long as I don't need to set my preference every time I open that popup and switch from one game to the next

> How about this layout? I prefer it because we won't have to deal with inconsistent, weird looking monospace fonts.

That wasts the same amount of space by just filling in the blanks through wrapping. If anything, monospace is as consistent as you can get with dates. I actually prefer a monospaced font when dealing with dates and columnar output. 
I even modified my local copy of the code to use the Meslo font for the date columns because I very much dislike a proportional display when dealing with numbers and dates.
No, what I meant was losing the connector line and circles around the icons and the blank line in between, a Timeline doesn't need to be literally visualized. The values themselves are explanatory enough to convey that.

That way the same data would use up about half the space as in below mock-up



> It will be harder to retroactively add other fields like last_updated, last_played, etc. because they are not static. 

I don't mean to use these fields permanently, but when this goes live, there is no historical data present. So, when you open this tab the first time when this feature is added, the first three entries for each game (`Added`, `Installed` and `Finished`) can be filled by using the values _existing at that moment_. All you need to do is check if there's any data yet, if not then read those three date values and any present version info for `Installed` and `Finished` _present at that time_. After that new entries can just use whatever source you're using now.

**Comment by FaceCrap** [comment]
changed my comment slightly, seeing as there's new commits since I first posted. Might not have been seen before making those commits

**Comment by FaceCrap** [comment]
> Yes, and to do that we'll have to add populate_timeline_with_old_events() function and run it before UI is initialized. 
> Not a big deal, but this little function will stay in the codebase forever 
> because it's impossible to tell when all users will switch to new version.

Not sure if you meant this function is now present (haven't looked at the code yet) but I wouldn't call that an issue, there's more one-time-use code in the codebase, like when a user switches from a version still using the json format for game data, and almost every time a database change was made. That almost can't be avoided if updating from an old(er) version requires to convert data without losing existing data. The change with game version tracking is I think the first time this didn't introduce one-time-use code.

Still, thank you very much for considering my suggestions.

**Comment by Willy-JL** [comment]
looking real nice, great work!

just a few things:
- tags could be both added and removed in one pass... checking just for different count isnt reliable. the code you have for differences is fine, just need to always check the differences, and then put the if based on length of differences. and not in an else case, just one after the other
- is there a particular reason as to why the nim_args is a string? just seems odd, but granted i dont remember how the enums i made work xD

and no there is no progress on the number of votes, though it should be fairly trivial. just been busy with flipper stuff as the project i was leading just got hijacked away from me, so im transferring to a new project

File: modules/api.py [hunk 1]
```diff
@@ -615,6 +616,26 @@ async def full_check(game: Game, version: str):
...
+        if game.type != type:
+            game.add_timeline_event(TimelineEventType.ChangedType, game.type.name, type.name)
+
+        if game.tags != tags:
+            if len(game.tags) < len(tags):
+                difference = [tag.text for tag in tags if tag not in game.tags]
+                game.add_timeline_event(TimelineEventType.TagsAdded, ", ".join(difference))
+            else:
+                difference = [tag.text for tag in game.tags if tag not in tags]
+                game.add_timeline_event(TimelineEventType.TagsRemoved, ", ".join(difference))
```
Comment (Willy-JL) [inline modules/api.py:631]: tags could be both added and removed in one pass... checking just for different count isnt reliable. the code you have for differences is fine, just need to always check the differences, and then put the if based on length of differences. and not in an else case, just one after the other

File: modules/structs.py [hunk 2]
```diff
@@ -477,6 +477,42 @@ def __post_init__(self):
...
+    ("ChangedVersion",   (7,  {"display": "Changed version",   "icon": "star",           "args_min": "2", "template": "Version changed from \"{}\" to \"{}\""})),
+    ("ChangedDeveloper", (8,  {"display": "Changed developer", "icon": "account",        "args_min": "2", "template": "Developer changed from \"{}\" to \"{}\""})),
+    ("ChangedType",      (9,  {"display": "Changed type",      "icon": "shape",          "args_min": "2", "template": "Type changed from \"{}\" to \"{}\""})),
+    ("TagsAdded",        (10, {"display": "Tags added",        "icon": "tag_plus",       "args_min": "1", "template": "Tags were added: {}"})),
+    ("TagsRemoved",      (11, {"display": "Tags removed",      "icon": "tag_minus",      "args_min": "1", "template": "Tags were removed: {}"})),
+    ("ScoreIncreased",   (12, {"display": "Score increased",   "icon": "thumb_up",       "args_min": "2", "template": "Forum score increased from {} to {}"})),
+    ("ScoreDecreased",   (13, {"display": "Score decreased",   "icon": "thumb_down",     "args_min": "2", "template": "Forum score decreased from {} to {}"})),
```
Comment (Willy-JL) [inline modules/structs.py:493]: is there a particular reason as to why the args_min is a string? just seems odd, but granted i dont remember how the enums i made work xD

**Comment by FaceCrap** [comment]
I hope this isn't too much to ask. I just thought about a possible new event candidate. 

Have you considered adding an event for Last Full Refresh?

**Comment by FaceCrap** [comment]
> > Have you considered adding an event for Last Full Refresh?
> 
> just to clarify, you mean firing an event every time when game is fully refreshed?

Yes, for those games getting refreshed in a weekly cycle so to speak, not just when updated.

**Comment by FaceCrap** [comment]
Ran into a bug...

If you select one game and force a full refresh, no problem, but if you select more than one game, then it crashes on the second game with the following trace stack.

```
Traceback (most recent call last):
  File "Y:\Code\VN\F95Checker-Timeline\main.py", line 98, in <module>
    main()
  File "Y:\Code\VN\F95Checker-Timeline\main.py", line 50, in main
    globals.gui.main_loop()
  File "Y:\Code\VN\F95Checker-Timeline\modules\gui.py", line 935, in main_loop
    self.draw_games_list()
  File "Y:\Code\VN\F95Checker-Timeline\modules\gui.py", line 3200, in draw_games_list
    self.handle_game_hitbox_events(game, drag_drop=True)
  File "Y:\Code\VN\F95Checker-Timeline\modules\gui.py", line 3029, in handle_game_hitbox_events
    self.draw_game_context_menu()
  File "Y:\Code\VN\F95Checker-Timeline\modules\gui.py", line 1629, in draw_game_context_menu
    self.draw_game_recheck_button(game, f"{icons.reload_alert} Full Recheck", selectable=True)
  File "Y:\Code\VN\F95Checker-Timeline\modules\gui.py", line 1542, in draw_game_recheck_button
    game.add_timeline_event(TimelineEventType.RecheckUserReq)
    ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'add_timeline_event'
```

**Comment by FaceCrap** [comment]
> it turns out [you introduced this bug yourself locally](https://github.com/Willy-JL/F95Checker/pull/138/files#diff-98b02e923daa2d3fec0b849fb13b77af895ebc55d43c287544877b9f64394658R1754).

🤣oh man, I've no idea how, But can't deny the link. Honestly didn't think to do a second pull.

**Comment by FaceCrap** [comment]
@Willy-JL 

image
I was wondering already where the 16 days in @r37r05p3C7's screenshot, of the event getting logged, came from, but this one took the cake. 😂

It's happens because in
```python
delta = dt.datetime.fromtimestamp(time.time()) - dt.datetime.fromtimestamp(game.last_full_check)
```
the `dt.datetime.fromtimestamp(game.last_full_check)` bit has the default epoch (`January 1, 1970, 00:00:00 (UTC`) value when adding a new game.
This causes a full recheck to happen twice...

The first time because I forced it to pick up the correct game details since I just added the game, 
and the second time due to the expiration interval getting triggered. 
To prevent duplicate logging (it's purely cosmetic, as it doesn't have any ill side-effect),
it might be an idea to initialize `game.last_full_check` also to `time.time()` when adding a new game,
since a user is going to force a full recheck anyway after adding.

**Comment by FaceCrap** [comment]
Just noticed that events don't get added for custom games, e.g. when you modify Version and Last Updated (only the date added is present).

**Comment by Willy-JL** [comment]
Been a bit out of the loop, sorry. Is anything left for this or can it be merged?

**Review by Willy-JL — APPROVED** [review]
:D

**Pull request merged**